infrastructure/cdn-in-a-box/enroller/enroller.go

```diff
@@ -713,6 +713,125 @@ func enrollServerCapability(toSession *session, r io.Reader) error {
 	return err
 }
 
+// enrollFederation takes a json file and creates a Federation object using the TO API.
+// It also assigns a Delivery Service, the CDN in a Box admin user, IPv4 resolvers,
+// and IPv6 resolvers to that Federation.
+func enrollFederation(toSession *session, r io.Reader) error {
+	dec := json.NewDecoder(r)
+	var federation tc.AllDeliveryServiceFederationsMapping
+	err := dec.Decode(&federation)
+	if err != nil {
+		log.Infof("error decoding Server Capability: %s\n", err)
+		return err
+	}
+	mapping := federation.Mappings[0]
+	var cdnFederation tc.CDNFederation
+	var cdnName string
+	{
+		xmlID := string(federation.DeliveryService)
+		deliveryServices, _, err := toSession.GetDeliveryServiceByXMLID(xmlID, nil)
+		if err != nil {
+			log.Infof("getting Delivery Service %s: %s", xmlID, err.Error())
+			return err
+		}
+		if len(deliveryServices) != 1 {
+			log.Infof("wanted 1 Delivery Service with XMLID %s but received %d Delivery Services", xmlID, len(deliveryServices))
+			return err
+		}
+		deliveryService := deliveryServices[0]
+		cdnName = *deliveryService.CDNName
+		cdnFederation = tc.CDNFederation{
+			CName: mapping.CName,
+			TTL:   mapping.TTL,
+		}
+		resp, _, err := toSession.CreateCDNFederation(cdnFederation, cdnName)
+		if err != nil {
+			log.Infof("creating CDN Federation: %s", err.Error())
+			return err
+		}
+		cdnFederation = resp.Response
+		if _, err = toSession.CreateFederationDeliveryServices(*resp.Response.ID, []int{*deliveryService.ID}, true); err != nil {
```

Comment:
there are a lot of places where dereferencing pointer values like this could be causing segfaults to crash the enroller. On one hand, other possible segfaults (like if `deliveryServices` has zero length above) that _are_ handled, so it seems like not crashing if TO (or the client) is broken is the intended behavior. But on the other hand it's difficult to say how the enroller could do anything useful if TO (or the client) is broken, which is the only way these could wind up null.

You can these problems or not, but I thought I'd just point it out.

Reply:
Generally as I'm going through the client methods in #5747 if any of them entail enroller changes I'm adding the necessary guards for API calls I have to modify, if that affects your decision at all. Which is to say, if you don't do it now I'll probably do it in a rebase at some point.

Reply:
I have have added a nil check where `enrollFederation()` dereferences the `UserName` field in e1b2105aeb.

As for the other dereferences, IMO, a nil check on on a Traffic Ops response struct field whose corresponding database column is part of that table's primary key or has a `NOT NULL` constraint is unnecessary.

Reply:
That being said, all the more power to you if you choose to add checks for them in #5747.

Reply:
It's certainly not as important as if this were a test - where we specifically shouldn't trust the software being tested - but for things like this it's really more important to just check user input, to be sure. But:

> As for the other dereferences, IMO, a nil check on on a Traffic Ops response struct field whose corresponding database column is part of that table's primary key or has a `NOT NULL` constraint is unnecessary.

In *my* opinion, a "nullable" field on a struct property that is `NOT NULL` or a primary key is unnecessary. That would solve this problem too.

Of course, that's not the place of this PR to change. Just food for thought.

As far as me changing them, it just offends my delicate sensibilities to have programs crash with memory access violations instead of printing an error and exiting non-zero normally. It's not really a security concern or anything, so I probably don't actually need to do anything, but I also probably still will.